parser/openfoodfacts_taxonomy_parser/parser.py

```diff
@@ -0,0 +1,336 @@
+        if not data["id"]:
+            data["id"] = id
+        else:
+            raise DuplicateIDError(line_number)
+        return data
+
+    def header_harvest(self, filename):
+        """to harvest the header (comment with #), it has its own function because some header has multiple blocks"""
+        h = 0
+        header = []
+        for _, line in self.file_iter(filename):
+            if not (line) or line[0] == "#":
+                header.append(line)
+            else:
+                break
+            h += 1
+
+        # we don't want to eat the comments of the next block and it remove the last separating line
+        for i in range(len(header)):
+            if header.pop():
+                h -= 1
+            else:
+                break
+
+        return header, h
+
+    def entry_end(self, line, data):
+        """Return True if the block ended"""
+        if "stopwords" in line or "synonyms" in line or not line:
+            # can be the end of an block or just 2 line separators,
+            # file_iter() always end with ''
+            if data["id"]:  # to be sure that it's an end
+                return True
+        return False
+
+    def remove_separating_line(self, data):
+        if data["preceding_lines"]:
+            if "synonyms" in data["id"]:
+                if "stopwords" in self.is_before:
+                    data["preceding_lines"].pop(0)
+            elif "stopwords" in data["id"]:
+                if "synonyms" in self.is_before:
+                    data["preceding_lines"].pop(0)
+            else:
+                data["preceding_lines"].pop(0)
+        return data
+
+    def harvest(self, filename):
+        """Transform data from file to dictionary"""
+        index_stopwords = 0
+        index_synonyms = 0
+        language_code_prefix = re.compile("[a-zA-Z][a-zA-Z][a-zA-Z]?:")
+        self.stopwords = dict()
+
+        # header
+        header, next_line = self.header_harvest(filename)
+        yield header
+        self.is_before = "__header__"
+
+        # the other entries
+        data = self.new_node_data()
+        for line_number, line in self.file_iter(filename, next_line):
+            # yield data if block ended
+            if self.entry_end(line, data):
+                data = self.remove_separating_line(data)
+                yield data  # another function will use this dictionary to create a node
+                self.is_before = data["id"]
+                data = self.new_node_data()
+
+            # harvest the line
+            if not (line) or line[0] == "#":
+                data["preceding_lines"].append(line)
+                if not data[
+                    "src_position"
+                ]:  # to get the position of the footer if it's not empty
+                    data["src_position"] = line_number + 1
+            else:
+                if (
+                    len(data) == 5 and not data["parent_tag"]
+                ):  # the beginning of the entry
```

Comment:
I find the `len(data) == 5` very obscure and fragile ! (it will break if I change `new_node_data`, while I won't be aware of it ! And I see many other way to unexpectedly break this !

For the footer you can compute it afterwards (you have the lenght of preceeding_line, so you can deduce where the foot start. Also it's not worth complexifying the code, just to get footer position right !

So I would remove the update of "src_position" above and here do:

```python
if not data["src_position"]:
    data["src_position"] = line_number
```

Even better, in `new_node_data`, you should put `src_position` to None and here, test with `is None`